### ISSUE TITLE
rancher2_bootstrap stops working when password is changed for admin user for rancher v2.6.0

### ISSUE THREAD

**Issue by abhi1693** [rancher/terraform-provider-rancher2#748]
I have updated our TF files when rancher v2.6.0 came out and the provider has been working just fine since that day. Today, due to our security policy a user updated the password from the UI and `rancher2_bootstrap` stopped working even though the updated password was passed to both `password` and `current_password` variables in the resource.

Here is a snippet of the relevant code

provider.tf
```
provider "rancher2" {
  alias = "bootstrap"

  api_url   = "https://${var.server_url}"
  bootstrap = true
  insecure  = true
}

provider "rancher2" {
  alias = "admin"

  api_url   = rancher2_bootstrap.admin.url
  token_key = rancher2_bootstrap.admin.token
  insecure  = true
}
```

main.tf
```
resource "rancher2_bootstrap" "admin" {
  provider   = rancher2.bootstrap
  depends_on = [null_resource.wait_for_rancher]

  current_password = var.admin_password
  password         = var.admin_password

  telemetry = false
  token_ttl = 900
}
```

Error after updating the password
```
module.rancher2.rancher2_bootstrap.admin: Refreshing state... [id=user-wrdhd]
2021-09-20T06:38:56.363-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:38:56 [INFO] Refreshing bootstrap
2021-09-20T06:38:56.363-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:38:56 [INFO] Doing login
2021-09-20T06:38:56.364-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:38:56 Getting from  https://rancher.local.com/ping
2021-09-20T06:38:56.461-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:38:56 Time to get req:  97  ms
2021-09-20T06:38:56.591-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:38:56 Getting from  https://rancher.local.com/ping
2021-09-20T06:38:56.702-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:38:56 Time to get req:  110  ms
2021-09-20T06:40:56.968-0700 [INFO]  backend/local: plan operation completed
╷
│ Error: [ERROR] Timeout login with admin user: Doing  user logging: error Unauthorized
│ 
│   with module.rancher2.rancher2_bootstrap.admin,
│   on ../../modules/rancher2/main.tf line 82, in resource "rancher2_bootstrap" "admin":
│   82: resource "rancher2_bootstrap" "admin" {
│ 
╵
Releasing state lock. This may take a few moments...
2021-09-20T06:40:57.594-0700 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-09-20T06:40:57.601-0700 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 pid=68604
2021-09-20T06:40:57.601-0700 [DEBUG] provider: plugin exited
```

Since I could not determine the root cause, we then reverted the password to the old one and haven't made any changes to the TF files yet. Here is the log for that

```
module.rancher2.rancher2_bootstrap.admin: Refreshing state... [id=user-wrdhd]
2021-09-20T06:45:31.766-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:31 [INFO] Refreshing bootstrap
2021-09-20T06:45:31.766-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:31 [INFO] Doing login
2021-09-20T06:45:31.766-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:31 Getting from  https://rancher.local.com/ping
2021-09-20T06:45:31.873-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:31 Time to get req:  106  ms
2021-09-20T06:45:32.011-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:32 Getting from  https://rancher.local.com/ping
2021-09-20T06:45:32.119-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:32 Time to get req:  107  ms
2021-09-20T06:45:32.482-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:32 Getting from  https://rancher.local.com/ping
2021-09-20T06:45:32.600-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:32 Time to get req:  118  ms
2021-09-20T06:45:33.769-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:33 [INFO] Connecting with user/pass
2021-09-20T06:45:34.023-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:34 [INFO] Cleaning up temp token
2021-09-20T06:45:34.073-0700 [WARN]  Provider "registry.terraform.io/rancher/rancher2" produced an unexpected new value for module.rancher2.rancher2_bootstrap.admin during refresh.
      - .temp_token: inconsistent values for sensitive attribute
      - .temp_token_id: was cty.StringVal(""), but now cty.StringVal("token-nqjws")
      - .token_update: was cty.False, but now cty.True
2021-09-20T06:45:34.079-0700 [WARN]  Provider "registry.terraform.io/rancher/rancher2" produced an invalid plan for module.rancher2.rancher2_bootstrap.admin, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .token_update: planned value cty.False for a non-computed attribute
      - .ui_default_landing: planned value cty.StringVal("ember") for a non-computed attribute
2021-09-20T06:45:34.081-0700 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-09-20T06:45:34.088-0700 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 pid=69774
2021-09-20T06:45:34.088-0700 [DEBUG] provider: plugin exited
2021-09-20T06:45:34.088-0700 [DEBUG] created provider logger: level=debug
2021-09-20T06:45:34.089-0700 [INFO]  provider: configuring client automatic mTLS
2021-09-20T06:45:34.215-0700 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 args=[.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0]
2021-09-20T06:45:34.215-0700 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 pid=69849
2021-09-20T06:45:34.215-0700 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0
2021-09-20T06:45:34.333-0700 [INFO]  provider.terraform-provider-rancher2_v1.20.0: configuring server automatic mTLS: timestamp=2021-09-20T06:45:34.333-0700
2021-09-20T06:45:34.484-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: plugin address: address=/tmp/plugin202027486 network=unix timestamp=2021-09-20T06:45:34.484-0700
2021-09-20T06:45:34.485-0700 [DEBUG] provider: using plugin: version=5
2021-09-20T06:45:34.665-0700 [DEBUG] created provider logger: level=debug
2021-09-20T06:45:34.665-0700 [INFO]  provider: configuring client automatic mTLS
2021-09-20T06:45:34.821-0700 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 args=[.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0]
2021-09-20T06:45:34.821-0700 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 pid=69862
2021-09-20T06:45:34.822-0700 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0
2021-09-20T06:45:34.940-0700 [INFO]  provider.terraform-provider-rancher2_v1.20.0: configuring server automatic mTLS: timestamp=2021-09-20T06:45:34.939-0700
2021-09-20T06:45:35.066-0700 [DEBUG] provider: using plugin: version=5
2021-09-20T06:45:35.066-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: plugin address: address=/tmp/plugin691709059 network=unix timestamp=2021-09-20T06:45:35.066-0700
2021-09-20T06:45:35.305-0700 [DEBUG] No provider meta schema returned
2021-09-20T06:45:35.321-0700 [DEBUG] No provider meta schema returned
2021-09-20T06:45:35.419-0700 [WARN]  ValidateProviderConfig from "module.rancher2.provider[\"registry.terraform.io/rancher/rancher2\"].admin" changed the config value, but that value is unused
2021-09-20T06:45:35.422-0700 [INFO]  ReferenceTransformer: reference not found: "rancher2_bootstrap.admin"
2021-09-20T06:45:35.422-0700 [INFO]  ReferenceTransformer: reference not found: "var.server_url"
2021-09-20T06:45:35.422-0700 [DEBUG] ReferenceTransformer: "module.rancher2.rancher2_setting.server_url" references: []
```

I have written a output.tf file capture the token generated from bootstrapping and passing it to other modules

output.tf
```
output "api_url" {
  value       = trimprefix(rancher2_bootstrap.admin.url, "https://")
  description = "API URL for the rancher server"
}

output "token" {
  value       = rancher2_bootstrap.admin.token
  description = "Token for the rancher server"
  sensitive   = true
}
```

I simply passed the `token` from the output.tf to other modules with the admin provider and here is an error log for one of them

```
module.monitoring.module.app.rancher2_namespace.ns: Refreshing state... [id=cattle-monitoring-system]
2021-09-20T06:45:35.444-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:35 [INFO] Refreshing Namespace ID cattle-monitoring-system
2021-09-20T06:45:35.444-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:35 Getting from  https://rancher.local.com/ping
2021-09-20T06:45:35.585-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:35 Time to get req:  140  ms
2021-09-20T06:45:35.595-0700 [DEBUG] provider.terraform-provider-rancher2_v1.20.0: 2021/09/20 06:45:35 Time to get req:  170  ms
2021-09-20T06:45:35.710-0700 [INFO]  backend/local: plan operation completed
╷
│ Error: Bad response statusCode [401]. Status [401 Unauthorized]. Body: [message=Unauthorized 401: must authenticate] from [https://rancher.local.com/v3]
│ 
│   with module.monitoring.module.app.rancher2_namespace.ns,
│   on ../../modules/rancher2-appv2/main.tf line 1, in resource "rancher2_namespace" "ns":
│    1: resource "rancher2_namespace" "ns" {
│ 
╵
╷
│ Error: Bad response statusCode [401]. Status [401 Unauthorized]. Body: [message=Unauthorized 401: must authenticate] from [https://rancher.local.com/v3]
│ 
│   with module.rancher2.rancher2_setting.server_url,
│   on ../../modules/rancher2/main.tf line 93, in resource "rancher2_setting" "server_url":
│   93: resource "rancher2_setting" "server_url" {
│ 
╵
Releasing state lock. This may take a few moments...
2021-09-20T06:45:36.387-0700 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-09-20T06:45:36.389-0700 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-09-20T06:45:36.392-0700 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 pid=69862
2021-09-20T06:45:36.392-0700 [DEBUG] provider: plugin exited
2021-09-20T06:45:36.393-0700 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/rancher/rancher2/1.20.0/linux_amd64/terraform-provider-rancher2_v1.20.0 pid=69849
2021-09-20T06:45:36.393-0700 [DEBUG] provider: plugin exited
```

Here is a screenshot from the rancher for the API token
![image](https://user-images.githubusercontent.com/5083532/134015083-27db5997-a659-4b2a-98ee-b0c2975bbec4.png)

@rawmind0 I have not made any changes in the TF files at all. I started getting all of these issues only after the admin password was changed.

Is it possible that the state file is using tokens from old runs which might be causing these issues and that the provider is not using the new information that I'm providing?

Another thing is that I have been able to reproduce this issue every single time consistently.

**Comment by abhi1693** [comment]
@rawmind0 As per your suggestion regarding the token being used first, I looked at the state file and it has the token info that does not exist on rancher. I think the `terraform plan` is referencing information that used to exist on rancher but does not anymore because when we updated the password we removed old API sessions.

Even though during the plan, a new token gets generated on Rancher, it's not used if it is referenced using the output from https://registry.terraform.io/providers/rancher/rancher2/latest/docs/resources/bootstrap#token

**Comment by abhi1693** [comment]
I'm stuck in a chicken and egg-like situation now. I'm unable to update the password because then bootstrapping stops working. If I revert back, the newly generated token is not used causing the plan to fail every time.

I have even tried to update the password using terraform and it fails again on bootstrapping during the planning stage.

@rawmind0  Your insight into this issue would be greatly appreciated.

**Comment by rawmind0** [comment]
@abhi1693 previous answer here, https://github.com/rancher/terraform-provider-rancher2/issues/738#issuecomment-922940166 . Have you checked you modules structure and var assignements?? Have you tested setting [token_update](https://registry.terraform.io/providers/rancher/rancher2/latest/docs/resources/bootstrap#token_update) argument to true?? This should force the token to be updated as explained at docs. 

I'm not able to reproduce your issue. Updating admin password from ui and then updating `password` and `current_password` tf arguments is working fine to me.

**Comment by abhi1693** [comment]
@rawmind0 

I have `token_update=true` as well but I still have the same issue. I have also tried to run terraform import on rancher2_bootstrap to download the updated information and it fails as it is not supported.

Can you try these steps

1. Using tfp, bootstrap rancher and setup any dependency like creating/updating cloud credential. Make sure that you pass the token only using the bootstrapped resources.
2. Update the password via UI and remove existing API sessions. 
![image](https://user-images.githubusercontent.com/5083532/134028793-9562e12e-2a40-4f38-bed6-15eb83b10242.png)
3. Then run your plan using the new password

The provider does create a new token but it is not used in any way for the dependencies. Unless the plan succeeds and the new changes can be applied, I don't believe this would work.

I have tried the above steps using a brand new setup and I'm stuck with the exact same issue.

If you are unable to still reproduce this issue, I'll just manually remove the objects from the state and re-import them after de-coupling the dependencies that I have which is causing these issues.

**Comment by rawmind0** [comment]
> The provider does create a new token but it is not used in any way for the dependencies. Unless the plan succeeds and the new changes can be applied, I don't believe this would work.

That's expected. Once you reverted admin to old password, have you tested just tf applying changes on `rancher2_bootstrap` , `terraform apply -target=rancher2_bootstrap.admin` . This plan should finish fine and update tfstate properly

**Comment by abhi1693** [comment]
@rawmind0 I'll try your solution. If that's expected, what's the best way to update the password in your opinion?

**Comment by rawmind0** [comment]
What it's expected at tf is `Unless the plan succeeds and the new changes can be applied, I don't believe this would work.` due to tfstate is not updated. 

If you are managing admin password from tfp, update it from the Rancher ui doesn't seems a good practice...Moreover when it's a key point for tfp to login to Rancher API. Starting from point where `rancher2_bootstrap` resource is working fine, just updating `password` argument and tf applying will do the job.

**Comment by abhi1693** [comment]
@rawmind0 It took a lot of manual work to revert the changes and bring back the state before the password was changed.

One last question, (this maybe out of scope for this provider), should I update the `bootstrapPassword` as well while changing the password?

**Comment by rawmind0** [comment]
@abhi1693 I'm sorry about manual work, but i'm glad to hear that the tfstate is fine again. 

About your question around `bootstrapPassword`, this setting is just for Rancher helm chart to set the initial Rancher admin password when you install a new Rancher v2.6.x. Once Rancher is installed, the update of this setting shouldn't be required, due to the bootstrap would be already done.

**Comment by abhi1693** [comment]
@rawmind0 Thanks for all your help. I'm closing this issue for now.
